core/src/main/java/org/opensearch/sql/expression/aggregation/AvgAggregator.java

```diff
@@ -47,18 +64,56 @@ public String toString() {
   /**
    * Average State.
    */
-  protected static class AvgState implements AggregationState {
-    private int count;
-    private double total;
+  protected abstract static class AvgState implements AggregationState {
+    protected int count;
+    protected double total;
 
     AvgState() {
       this.count = 0;
       this.total = 0d;
     }
 
+    @Override
+    public abstract ExprValue result();
+
+    protected AvgState iterate(ExprValue value) {
+      count++;
+      return this;
+    }
+  }
+
+  protected static class DoubleAvgState extends AvgState {
     @Override
     public ExprValue result() {
-      return count == 0 ? ExprNullValue.of() : ExprValueUtils.doubleValue(total / count);
+      if (count == 0) {
+        return ExprNullValue.of();
+      }
+      return ExprValueUtils.doubleValue(total / count);
+    }
+
+    @Override
+    protected AvgState iterate(ExprValue value) {
+      total += ExprValueUtils.getDoubleValue(value);
+      return super.iterate(value);
+    }
+  }
+
+  @RequiredArgsConstructor
+  protected static class DateTimeAvgState extends AvgState {
+    private final ExprCoreType dataType;
+
+    @Override
+    public ExprValue result() {
+      if (count == 0) {
+        return ExprNullValue.of();
+      }
+      return ExprValueUtils.convertEpochMilliToDateTimeType(Math.round(total / count), dataType);
+    }
+
+    @Override
+    protected AvgState iterate(ExprValue value) {
+      total += ExprValueUtils.extractEpochMilliFromAnyDateTimeType(value);
+      return super.iterate(value);
```

Comment:
An improvement I'm thinking in https://github.com/opensearch-project/sql/pull/1045 recently is can we build our physical operator based on expression system and avoid using native Java as much as possible. If we can treat our expression system as a higher level language than Java, we can avoid many logic already implemented in our expr value and functions, such as explicit/implicit conversion, arithmetic and lots of other operations.

For example, can we do it in this way:

```
class AvgState:
   total = doubleValue(0.0)
   count = integerValue(0)

  def iterate(value: ExprValue):
    count = DSL.add(count, integerValue(1))
    total = DSL.add(total, value) // specific ExprValue is responsible for converting itself to ExprDoubleValue
```

Reply:
Thanks for the idea. I'll have a look and check how can I update this after #1045.

Reply:
Sure. Just some thought. I'm working on https://github.com/opensearch-project/sql/pull/1045 and checking if I can re-publish it to 2.x once done. Thanks!

Reply:
`result` function should return an `ExprValue`:
https://github.com/opensearch-project/sql/blob/e280866ad6a248e852528326ecf568c95b4958a2/core/src/main/java/org/opensearch/sql/expression/aggregation/AggregationState.java#L15-L19

Are you planning to update all aggregators to return `Expression` instead of `ExprValue`?
Without changing the interface, implementation looks like
```
def result():
  return DSL.divide(total, count).valueOf();
```

Is ok for now? I feel that we are loosing all benefits of `DSL` because we switch down to java.

Reply:
Sorry I mixed the Expression and ExprValue in the sample code above. Here is what I was thinking about. Let me know how you feel implement our physical operator in this way.

```
class AvgState:
   total = doubleValue(0.0)
   count = integerValue(0)

  def iterate(value: ExprValue):
    count = DSL.add(DSL.literal(count), DSL.literal(1)).valueOf()
    total = DSL.add(DSL.literal(total), DSL.castToDouble(DSL.literal(value))).valueOf()

  def result():
    return DSL.divide(DSL.literal(total), DSL.literal(count)).valueOf()
```

Reply:
Updated in c7734442e.

Reply:
> Are you planning to update all aggregators to return `Expression` instead of `ExprValue`?

This will make all in-memory aggregation to use DSL. Probably this bring a benefit for Analyzer and/or Optimizer in future.
Please, create a follow-up task if you feel we need to do this.